clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/view/provider/KubernetesV2ClusterProvider.java

```diff
@@ -140,21 +169,88 @@ public boolean supportsMinimalClusters() {
   }
 
   private Set<KubernetesV2Cluster> translateClusters(Collection<CacheData> clusterData) {
-    // TODO(lwander) resolve server group & lb relationships
     return clusterData.stream().map(this::translateCluster).filter(Objects::nonNull).collect(Collectors.toSet());
   }
 
+  private Set<KubernetesV2Cluster> translateClustersWithRelationships(Collection<CacheData> clusterData) {
+    // TODO(lwander) possible optimization: store lb relationships in cluster object to cut down on number of loads here.
+    List<CacheData> serverGroupData = kindMap.translateSpinnakerKind(SERVER_GROUP)
+        .stream()
+        .map(kind -> cacheUtils.loadRelationshipsFromCache(clusterData, kind.toString()))
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+
+    List<CacheData> loadBalancerData = kindMap.translateSpinnakerKind(LOAD_BALANCER)
+        .stream()
+        .map(kind -> cacheUtils.loadRelationshipsFromCache(serverGroupData, kind.toString()))
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+
+    Map<String, List<CacheData>> clusterToServerGroups = new HashMap<>();
+    for (CacheData serverGroupDatum : serverGroupData) {
+      Collection<String> clusterKeys = serverGroupDatum.getRelationships().get(CLUSTER.toString());
+      if (clusterKeys == null || clusterKeys.size() != 1) {
+        log.warn("Malformed cache, server group stored without cluster");
+        continue;
+      }
+
+      String clusterKey = clusterKeys.iterator().next();
+      List<CacheData> storedData = clusterToServerGroups.getOrDefault(clusterKey, new ArrayList<>());
+      storedData.add(serverGroupDatum);
+      clusterToServerGroups.put(clusterKey, storedData);
+    }
+
+    Map<String, List<CacheData>> serverGroupToLoadBalancers = new HashMap<>();
+    for (CacheData loadBalancerDatum : loadBalancerData) {
+      List<String> serverGroupKeys = kindMap.translateSpinnakerKind(LOAD_BALANCER)
+          .stream()
+          .map(t -> loadBalancerDatum.getRelationships().get(t.toString()))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+
+      for (String serverGroupKey : serverGroupKeys) {
+        List<CacheData> storedData = serverGroupToLoadBalancers.getOrDefault(serverGroupKey, new ArrayList<>());
+        storedData.add(loadBalancerDatum);
+        serverGroupToLoadBalancers.put(serverGroupKey, storedData);
+      }
+    }
+
+    Set<KubernetesV2Cluster> result = new HashSet<>();
+    for (CacheData clusterDatum : clusterData) {
```

Comment:
Could use `forEach` for these loops.

Reply:
I'm not a big fan of big, multi-line lambdas. They always feel kind-of forced & break some of the elegance of a few chained stream processing calls. Since this captures a fair bit of state, pulling this into a helper felt like overkill.